### ISSUE TITLE
Arguments.Values property or Arguments.CanResolve method

### ISSUE THREAD

**Issue by jnm2** [castleproject/Windsor#470]
When upgrading a codebase to Castle.Windsor 5, I ended up having to change the `if` check to the following:

```cs
registration.DynamicParameters((k, arguments) =>
{
    if (!arguments.Any(arg => entityType.IsInstanceOfType(arg.Value)))
    {
        arguments["foo"] = Activator.CreateInstance(entityType);
    }
});
```

The question I'm trying to answer is, will Castle.Windsor resolve the parameter named `foo` whose type is `entityType` from the `arguments` collection, or do I need to step in and provide a default?

It occurred to me that maybe I should try this, but I'm afraid it's not as closely aligned with what Windsor will actually do with `arguments`. For example, the type might not be an exact match:

```cs
if (!(arguments.Contains("foo") || arguments.Contains(entityType))`
{
```

Is `arguments.Any(arg => entityType.IsInstanceOfType(arg.Value))` the best way to do what I want to do?

- If so, would you consider exposing the `.Values` dictionary property to enable this syntax, closer to the syntax I used in v4?
  ```cs
  if (!arguments.Values.Any(entityType.IsInstanceOfType))`
  {
      arguments["foo"] = Activator.CreateInstance(entityType);
  }
  ```

- If it's more subtle than my `.Any(IsInstanceOfType)`, would you consider adding something like a `.CanResolve` helper so that I can ask the question without making any assumptions?
  ```cs
  if (!arguments.CanResolve(entityType, "foo"))`
  {
      arguments["foo"] = Activator.CreateInstance(entityType);
  }
  ```

**Comment by jonorossi** [comment]
I'm not completely across what you are doing because I can't see the code before the example, however I'm guessing you are sort of looking for a `ContainsValue` like `ContainsKey`.

I think we wanted to avoid exposing the internal values collection because it is an implementation detail that both named and typed arguments are stored in the one collection at the moment.

The CanResolve method idea sounds interesting. What is your preference for the API?

**Comment by jnm2** [comment]
This applies to certain view models that allow the user to edit a new or existing entity. Either a typed factory with a `Resolve(SomeEntity entity)` method was used to create the view model and parameterize the instance of the entity to be created or edited, or a parameterless `Resolve` method was used.

In the latter case, this code steps in and passes `Activator.CreateInstance(entityType)` since no entity instance was provided, since otherwise Windsor would fail to instantiate the view model.

This is old code and I'm open to new ways of doing things. The drawback to an optional parameter is that it's the first constructor parameter with a large number of non-optional parameters following it. If I move it to the end, it's harder to notice its significance after all the noise of the injected services. (Perhaps I should look for a way to address that, too.)

> I think we wanted to avoid exposing the internal values collection because it is an implementation detail that both named and typed arguments are stored in the one collection at the moment.

Would exposing an `IReadOnlyCollection<object> Values` property be exposing an implementation detail? You already allow iterating over all arguments as `IEnumerable<KeyValuePair<object, object>>`. Regardless of the internal implementation, it seems reasonable that each argument has a value and that you might want all the values without all the keys. The implementation right now would be very easy, but it would still seem trivial if you were using two collections internally.

My concern is whether considering just the values is something I should be doing in this scenario.

> The CanResolve method idea sounds interesting. What is your preference for the API?

I'm not sure. I'd like a way to ask, "Will Castle Windsor be able to resolve a parameter named `x` of type `y` from this Arguments collection?" So that if the answer is no, I can add an argument with that name and type. Is `CanResolveParameter` too specific or the wrong terminology? `CanResolveDependency`?

**Comment by jonorossi** [comment]
> Would exposing an `IReadOnlyCollection<object> Values` property be exposing an implementation detail? Windsor already allows iterating over all arguments as `IEnumerable<KeyValuePair<object, object>>`. Regardless of the internal implementation, it seems reasonable that each argument has a value and that you might want all the values without all the keys. The implementation right now would be very easy, but it would still seem trivial if you were using two collections internally.

True. Based on your reply, I see no harm in adding a `Values` property.

> I'm not sure. I'd like a way to ask, "Will Castle Windsor be able to resolve a parameter named `x` of type `y` from this Arguments collection?" So that if the answer is no, I can add an argument with that name and type. Is `CanResolveParameter` too specific or the wrong terminology? `CanResolveDependency`?

I think anything with the name "resolve" sounds too important like it is doing important stuff, since it is basically a dumb K/V collection. Could maybe add `TryGet` since the actual Windsor logic isn't inside this collection, but the indexer already does this and returns null if it doesn't exist.

There are probably endless ways to implement this, I'm happy if you want to add the `Values` property and move on.
